src/java_bytecode/java_bytecode_language.cpp

```diff
@@ -195,6 +195,52 @@ bool java_bytecode_languaget::parse(
   return false;
 }
 
+/// Infer fields that must exist on opaque types from field accesses against
+/// them. Note that we don't yet try to infer inheritence between opaque types
+/// here, so we may introduce the same field onto a type and its ancestor
+/// without realising that is in fact the same field, inherited from that
+/// ancestor.
+/// \param parse_tree: class parse tree
+/// \param symbol_table: global symbol table
+static void infer_opaque_type_fields(
+  const java_bytecode_parse_treet &parse_tree,
+  symbol_tablet &symbol_table)
+{
+  for(const auto &method : parse_tree.parsed_class.methods)
+  {
+    for(const auto &instruction : method.instructions)
+    {
+      if(instruction.statement == "getfield" ||
+         instruction.statement == "putfield")
+      {
+        const exprt &fieldref = instruction.args[0];
+        const symbolt *class_symbol =
+          symbol_table.lookup(fieldref.get(ID_class));
+        INVARIANT(
+          class_symbol != nullptr, "all field types should have been loaded");
+
+        if(class_symbol->type.get_bool(ID_incomplete_class))
+        {
+          // Accessing a field of an incomplete (opaque) type. Add it if not
+          // already present.
+          const irep_idt &component_name = fieldref.get(ID_component_name);
+          if(!to_struct_type(class_symbol->type).has_component(component_name))
+          {
+            symbolt &writable_class_symbol = symbol_table.get_writeable_ref(
+              fieldref.get(ID_class));
+            auto &add_to_components =
+              to_struct_type(writable_class_symbol.type).components();
+            add_to_components.push_back(
+              struct_typet::componentt(component_name, fieldref.type()));
+            add_to_components.back().set_base_name(component_name);
+            add_to_components.back().set_pretty_name(component_name);
```

Comment:
I don't know if required here, but it might be sensible to put a default access using `set_access`?